src/components/game/Topology.js

```diff
@@ -0,0 +1,386 @@
+import { FIELD_SIZE } from './const';
+import { Draw } from './Draw';
+import {logPlugin} from "@babel/preset-env/lib/debug";
+
+export class Topology {
+  constructor(param) {
+    this.offsetX = param.offsetX
+    this.offsetY = param.offsetY
+    // ссылка на объект реализации
+    this.drawer = new Draw({
+      offsetX: this.offsetX,
+      offsetY: this.offsetY
+    })
+    // для того, чтобы не показывать корабли бота
+    this.secret = param.secret || false // БОТ
+
+    this.sheeps = []
+    this.checks = []
+    this.injuries = []
+    this.kills = []
+    this.last = {}
+    this.score = {}
+  }
+
+  // добавление кораблей
+  // принимает какое-то кол-во кораблей и добавляет их в массив sheeps
+  addSheeps(...sheeps) {
+    // проверяем не был ли ранее добавлен данный корабль,
+    // если не был, то добавляем в sheeps
+    for (const sheep of sheeps) {
+      if (!this.sheeps.includes(sheep)) {
+        this.sheeps.push(sheep)
+      }
+    }
+    return this
+  }
+
+  // добавление выстрелов
+  // принимает какое-то кол-во выстрелов и добавляет их в массив checks
+  addChecks(...checks) {
+    // проверяем не был ли ранее добавлен данный выстрел,
+    // если не был, то добавляем в checks
+    for (const check of checks) {
+      if (!this.checks.includes(check)) {
+        this.checks.push(check)
+      }
+    }
+    return this
+  }
+
+  // проверяем не был ли ранее добавлен данный выстрел
+  isChecked(point) {
+    const flag = this.checks.find((check) => check.x === point.x && check.y === point.y)
+    if (flag) {
+      return true
+    }
+    return false
+  }
+
+  // добавление последнего хода
+  addThelast(last) {
+    this.last = last
+    return this
+  }
+
+  getScore(x, y) {
+    this.score = {
+      x,
+      y,
+      count: this.kills.length
+    }
+  }
+
+  // метод абстракции, делегирует работу связанному методу реализации
+  draw(context) {
+    this.drawer.drawFields(context)
+
+    // Отрисовка или не отрисовка кораблей
+    // БОТ
+    if (!this.secret) {
+      for (const sheep of this.sheeps) {
+        this.drawer.drawSheep(context, sheep)
+      }
+    }
+
+    for (const check of this.checks) {
+      this.drawer.drawCheck(context, check)
+    }
+
+    for (const injury of this.injuries) {
+      this.drawer.drawInjury(context, injury)
+    }
+
+    for (const sheep of this.kills) {
+      this.drawer.drawCheckAroundKills(context, sheep, this.checks)
+    }
+
+    this.drawer.drawLast(context, this.last)
+
+    this.drawer.drawScore(context, this.score)
+
+    return this
+  }
+
+  // положение мышки. true, если в области поля морского боя
+  // false, если на буквах, цифрах или вне поля
+  isPointUnder(point) {
+    if (
+      point.x < this.offsetX + FIELD_SIZE
+            || point.x > this.offsetX + 11 * FIELD_SIZE
+            || point.y < this.offsetY + FIELD_SIZE
+            || point.y > this.offsetY + 11 * FIELD_SIZE
+    ) {
+      return false
+    }
+    return true
```

Comment:
❗️
Много где по коду лишние ветки условий. Давай их дропнем 
```suggestion
    return !(
      point.x < this.offsetX + FIELD_SIZE
            || point.x > this.offsetX + 11 * FIELD_SIZE
            || point.y < this.offsetY + FIELD_SIZE
            || point.y > this.offsetY + 11 * FIELD_SIZE
    ) 
```

Reply:
поправил! пройдусь проверю!